File: 04.03.特定深度节点链表/04.03.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func listOfDepth(tree *TreeNode) []*ListNode {
	res := make([]*ListNode, 0)
	if tree == nil {
		return res
	}

	que := make([]*TreeNode, 0)
	que = append(que, tree)

	for len(que) > 0 {
		var head, prev *ListNode
		size := len(que)

		for i := 0; i < size; i++ {
			cur := que[0]
			que = que[1:]

			node := &ListNode{Val: cur.Val}

			if head == nil {
				head = node
			} else {
				prev.Next = node
			}
			prev = node

			if cur.Left != nil {
				que = append(que, cur.Left)
			}
			if cur.Right != nil {
				que = append(que, cur.Right)
			}
		}

		res = append(res, head)
	}

	return res
}

File: 04.03.特定深度节点链表/04.03.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> listOfDepth(TreeNode* tree) {
        vector<ListNode*> res;
        if(tree == nullptr) {
            return res;
        }

        queue<TreeNode*> q;
        q.push(tree);

        while(!q.empty()) {
            ListNode* head = nullptr;
            ListNode* prev = nullptr;

            int size = q.size();
            for(int i = 0; i < size; ++i) {
                TreeNode* cur = q.front();
                q.pop();
                ListNode* node = new ListNode(cur->val);

                if(head == nullptr) {
                    head = node;
                } else {
                    prev->next = node;
                }
                prev = node;

                if(cur->left != nullptr) {
                    q.push(cur->left);
                }
                if(cur->right != nullptr) {
                    q.push(cur->right);
                }
            }
            
            res.push_back(head);
        }

        return res;
    }
};